# Patch to fix cofiguring and building with Ruby 1.9 and Lua 5.2
# See https://gitweb.gentoo.org/repo/gentoo.git/tree/www-client/elinks/files/elinks-0.12_pre5-ruby-1.9.patch
diff -urNad elinks-0.12pre6-orig/config/m4/ruby.m4 elinks-0.12pre6/config/m4/ruby.m4
--- elinks-0.12pre6-orig/config/m4/ruby.m4	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/config/m4/ruby.m4	2016-10-25 18:41:10.000000000 +0300
@@ -32,36 +32,40 @@
 	if test "$CONFIG_SCRIPTING_RUBY" != "no"; then
 
 		AC_MSG_CHECKING(Ruby version)
-		if $CONFIG_SCRIPTING_RUBY -e 'exit((VERSION or RUBY_VERSION) >= "1.6.0")' >/dev/null 2>/dev/null; then
+		if $CONFIG_SCRIPTING_RUBY -e 'exit((VERSION rescue RUBY_VERSION) >= "1.6.0")' >/dev/null 2>/dev/null; then
 			ruby_version=`$CONFIG_SCRIPTING_RUBY -e 'puts "#{VERSION rescue RUBY_VERSION}"'`
 			AC_MSG_RESULT($ruby_version)
 
 			AC_MSG_CHECKING(for Ruby header files)
-			rubyhdrdir=`$CONFIG_SCRIPTING_RUBY -r mkmf -e 'print Config::CONFIG[["archdir"]] || $hdrdir' 2>/dev/null`
+			rubyhdrdir=`$CONFIG_SCRIPTING_RUBY -r mkmf -e 'print RbConfig::CONFIG[["rubyhdrdir"]] || RbConfig::CONFIG[["archdir"]] || $hdrdir' 2>/dev/null`
 
 			if test "X$rubyhdrdir" != "X"; then
 				AC_MSG_RESULT($rubyhdrdir)
 				RUBY_CFLAGS="-I$rubyhdrdir"
-				rubylibs=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print Config::CONFIG[["LIBS"]]'`
+				rubyarch=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print RbConfig::CONFIG[["arch"]]'`
+				if test -d "$rubyhdrdir/$rubyarch"; then
+					RUBY_CFLAGS="$RUBY_CFLAGS -I$rubyhdrdir/$rubyarch"
+				fi
+				rubylibs=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print RbConfig::CONFIG[["LIBS"]]'`
 
 				if test "X$rubylibs" != "X"; then
 					RUBY_LIBS="$rubylibs"
 				fi
 
-				librubyarg=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print Config.expand(Config::CONFIG[["LIBRUBYARG"]])'`
+				librubyarg=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print RbConfig.expand(RbConfig::CONFIG[["LIBRUBYARG"]])'`
 
 				if test -f "$rubyhdrdir/$librubyarg"; then
 					librubyarg="$rubyhdrdir/$librubyarg"
 
 				else
-					rubylibdir=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print Config.expand(Config::CONFIG[["libdir"]])'`
+					rubylibdir=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print RbConfig.expand(RbConfig::CONFIG[["libdir"]])'`
 					if test -f "$rubylibdir/$librubyarg"; then
 						librubyarg="$rubylibdir/$librubyarg"
 					elif test "$librubyarg" = "libruby.a"; then
 						dnl required on Mac OS 10.3 where libruby.a doesn't exist
 						librubyarg="-lruby"
 					else
-						librubyarg=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e "print '$librubyarg'.gsub(/-L\./, %'-L#{Config.expand(Config::CONFIG[\"libdir\"])}')"`
+						librubyarg=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e "print '$librubyarg'.gsub(/-L\./, %'-L#{RbConfig.expand(RbConfig::CONFIG[\"libdir\"])}')"`
 					fi
 				fi
 
@@ -69,7 +73,7 @@
 					RUBY_LIBS="$librubyarg $RUBY_LIBS"
 				fi
 
-				rubyldflags=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print Config::CONFIG[["LDFLAGS"]]'`
+				rubyldflags=`$CONFIG_SCRIPTING_RUBY -r rbconfig -e 'print RbConfig::CONFIG[["LDFLAGS"]]'`
 				if test "X$rubyldflags" != "X"; then
 					LDFLAGS="$rubyldflags $LDFLAGS"
 				fi
@@ -86,6 +90,15 @@
 			AC_MSG_RESULT(too old; need Ruby version 1.6.0 or later)
 		fi
 	fi
+	if test "$CONFIG_SCRIPTING_RUBY" = "yes"; then
+		AC_MSG_CHECKING([for rb_errinfo])
+		AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <ruby.h>]], [[rb_errinfo();]])],have_rb_errinfo="yes",have_rb_errinfo="no")
+		AC_MSG_RESULT($have_rb_errinfo)
+		if test "$have_rb_errinfo" = "yes"; then
+			AC_DEFINE([HAVE_RB_ERRINFO], [1],
+				[Define to 1 if you have the `rb_errinfo' function.])
+		fi
+	fi
 fi
 
 EL_RESTORE_FLAGS
# # For Slackware I do not provide lualib, so remove it and use Lua 5.2, fix configtest.c for Lua 5.2
diff -urNad elinks-0.12pre6-orig/configure.in elinks-0.12pre6/configure.in
--- elinks-0.12pre6-orig/configure.in	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/configure.in	2016-10-25 19:44:23.000000000 +0300
@@ -834,10 +834,10 @@
 		withval="";
 	fi
 	for luadir in "$withval" "" /usr /usr/local; do
-		for suffix in "" 50; do
+		for suffix in "" 52; do
 			if test "$cf_result" = no && ( test -f "$luadir/include/lua.h" || \
 			   test -f "$luadir/include/lua$suffix/lua.h" ) ; then
-				LUA_LIBS="-L$luadir/lib -llua$suffix -llualib$suffix -lm"
+				LUA_LIBS="-L$luadir/lib -llua -lm"
 				LUA_CFLAGS="-I$luadir/include -I$luadir/include/lua$suffix"
 
 				LIBS="$LUA_LIBS $LIBS_X"
@@ -846,14 +846,8 @@
 
 				# Check that it is a compatible Lua version
 				AC_LINK_IFELSE([AC_LANG_PROGRAM([[	#include <lua.h>
-						#include <lualib.h>]], [[	lua_State *L = lua_open();
-						luaopen_base(L);
-						luaopen_table(L);
-						luaopen_io(L);
-						luaopen_string(L);
-						luaopen_math(L);
-						lua_pushboolean(L, 1);
-						lua_close(L);]])],[cf_result=yes],[cf_result=no])
+						#include <lualib.h>]], [[	lua_State *L = luaL_newstate();
+						luaL_openlibs(L);]])],[cf_result=yes],[cf_result=no])
 			fi
 		done
 	done
# Fix compilation error due to API changes in Lua 5.1 and 5.2
diff -urNad elinks-0.12pre6-orig/src/scripting/lua/core.c elinks-0.12pre6/src/scripting/lua/core.c
--- elinks-0.12pre6-orig/src/scripting/lua/core.c	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/src/scripting/lua/core.c	2016-10-25 20:01:57.000000000 +0300
@@ -658,7 +658,7 @@
 	if (file_can_read(file)) {
 		int oldtop = lua_gettop(S);
 
-		if (lua_dofile(S, file) != 0)
+		if (luaL_dofile(S, file) != 0)
 			sleep(3); /* Let some time to see error messages. */
 		lua_settop(S, oldtop);
 	}
@@ -669,13 +669,9 @@
 void
 init_lua(struct module *module)
 {
-	L = lua_open();
+	L = luaL_newstate();
 
-	luaopen_base(L);
-	luaopen_table(L);
-	luaopen_io(L);
-	luaopen_string(L);
-	luaopen_math(L);
+	luaL_openlibs(L);
 
 	lua_register(L, LUA_ALERT, l_alert);
 	lua_register(L, "current_url", l_current_url);
@@ -780,7 +776,7 @@
 		int oldtop = lua_gettop(L);
 
 		if (prepare_lua(ses) == 0) {
-			lua_dostring(L, expr);
+			luaL_dostring(L, expr);
 			lua_settop(L, oldtop);
 			finish_lua();
 		}
diff -urNad elinks-0.12pre6-orig/src/scripting/lua/hooks.c elinks-0.12pre6/src/scripting/lua/hooks.c
--- elinks-0.12pre6-orig/src/scripting/lua/hooks.c	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/src/scripting/lua/hooks.c	2016-10-25 20:03:20.000000000 +0300
@@ -140,7 +140,7 @@
 	if (err) return EVENT_HOOK_STATUS_NEXT;
 
 	if (lua_isstring(L, -1)) {
-		int len = lua_strlen(L, -1);
+		int len = lua_rawlen(L, -1);
 
 		add_fragment(cached, 0, (unsigned char *) lua_tostring(L, -1), len);
 		normalize_cache_entry(cached, len);
@@ -200,7 +200,7 @@
 script_hook_quit(va_list ap, void *data)
 {
 	if (!prepare_lua(NULL)) {
-		lua_dostring(lua_state, "if quit_hook then quit_hook() end");
+		luaL_dostring(lua_state, "if quit_hook then quit_hook() end");
 		finish_lua();
 	}
 
diff -urNad elinks-0.12pre6-orig/src/scripting/ruby/core.c elinks-0.12pre6/src/scripting/ruby/core.c
--- elinks-0.12pre6-orig/src/scripting/ruby/core.c	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/src/scripting/ruby/core.c	2016-10-25 18:41:10.000000000 +0300
@@ -76,10 +76,10 @@
 		break;
 	case TAG_RAISE:
 	case TAG_FATAL:
-		eclass = CLASS_OF(ruby_errinfo);
-		einfo = rb_obj_as_string(ruby_errinfo);
+		eclass = CLASS_OF(RB_ERRINFO);
+		einfo = rb_obj_as_string(RB_ERRINFO);
 
-		if (eclass == rb_eRuntimeError && RSTRING(einfo)->len == 0) {
+		if (eclass == rb_eRuntimeError && RSTRING_LEN(einfo) == 0) {
 			msg = "unhandled exception";
 
 		} else {
@@ -88,7 +88,7 @@
 
 			epath = rb_class_path(eclass);
 			snprintf(buff, MAX_STR_LEN, "%s: %s",
-				RSTRING(epath)->ptr, RSTRING(einfo)->ptr);
+				RSTRING_PTR(epath), RSTRING_PTR(einfo));
 
 			p = strchr(buff, '\n');
 			if (p) *p = '\0';
@@ -115,7 +115,7 @@
 	unsigned char *message, *line_end;
 
 	str = rb_obj_as_string(str);
-	message = memacpy(RSTRING(str)->ptr, RSTRING(str)->len);
+	message = memacpy(RSTRING_PTR(str), RSTRING_LEN(str));
 	if (!message) return Qnil;
 
 	line_end = strchr(message, '\n');
@@ -162,8 +162,8 @@
 		 * the inspect() method, which adds quotes to the strings, so
 		 * gently ignore them. */
 
-		ptr = RSTRING(substr)->ptr;
-		len = RSTRING(substr)->len;
+		ptr = RSTRING_PTR(substr);
+		len = RSTRING_LEN(substr);
 
 		if (*ptr == '"')
 			ptr++, len--;
diff -urNad elinks-0.12pre6-orig/src/scripting/ruby/core.h elinks-0.12pre6/src/scripting/ruby/core.h
--- elinks-0.12pre6-orig/src/scripting/ruby/core.h	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/src/scripting/ruby/core.h	2016-10-25 18:41:11.000000000 +0300
@@ -7,6 +7,20 @@
 
 #include <ruby.h>	/* for VALUE */
 
+#ifndef RSTRING_LEN
+#define RSTRING_LEN(string) (RSTRING(string)->len)
+#endif
+
+#ifndef RSTRING_PTR
+#define RSTRING_PTR(string) (RSTRING(string)->ptr)
+#endif
+
+#ifdef HAVE_RB_ERRINFO
+#define RB_ERRINFO (rb_errinfo())
+#else
+#define RB_ERRINFO (ruby_errinfo)
+#endif
+
 VALUE erb_module;
 
 void alert_ruby_error(struct session *ses, unsigned char *msg);
diff -urNad elinks-0.12pre6-orig/src/scripting/ruby/hooks.c elinks-0.12pre6/src/scripting/ruby/hooks.c
--- elinks-0.12pre6-orig/src/scripting/ruby/hooks.c	2012-10-28 14:57:15.000000000 +0200
+++ elinks-0.12pre6/src/scripting/ruby/hooks.c	2016-10-25 18:41:11.000000000 +0300
@@ -83,7 +83,7 @@
 	{
 		unsigned char *new_url;
 
-		new_url = memacpy(RSTRING(result)->ptr, RSTRING(result)->len);
+		new_url = memacpy(RSTRING_PTR(result), RSTRING_LEN(result));
 		if (new_url) {
 			mem_free_set(url, new_url);
 		}
@@ -126,7 +126,7 @@
 	{
 		unsigned char *new_url;
 
-		new_url = memacpy(RSTRING(result)->ptr, RSTRING(result)->len);
+		new_url = memacpy(RSTRING_PTR(result), RSTRING_LEN(result));
 		if (new_url) {
 			mem_free_set(url, new_url);
 		}
@@ -170,9 +170,9 @@
 	switch (rb_type(result)) {
 	case T_STRING:
 	{
-		int len = RSTRING(result)->len;
+		int len = RSTRING_LEN(result);
 
-		add_fragment(cached, 0, RSTRING(result)->ptr, len);
+		add_fragment(cached, 0, RSTRING_PTR(result), len);
 		normalize_cache_entry(cached, len);
 
 		break;
@@ -216,7 +216,7 @@
 	{
 		unsigned char *proxy;
 
-		proxy = memacpy(RSTRING(result)->ptr, RSTRING(result)->len);
+		proxy = memacpy(RSTRING_PTR(result), RSTRING_LEN(result));
 		if (proxy) {
 			mem_free_set(new_proxy_url, proxy);
 		}
